### PR TITLE
Create object without setting PK is not allowed

### PR DESCRIPTION
**Pull request by beeender**
* createObject must be called with PK value when creating a object with
  PK defined.
* Creating objects from JSON must have have corresponding PK defined in
  the JSON object.

Known issue:
The default values for creating object from JSONStream will be differnt
from those created by createObject. Default values from default
constructor VS default values from core. This has to be addressed by #777

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@
   - `RealmIOExcpetion` has been removed and replaced by `RealmFileException`.
 * Removed `RealmConfiguration.Builder(Context, File)` and `RealmConfiguration.Builder(File)` constructors.
 * `RealmConfiguration.Builder.assetFile(Context, String)` has been renamed to `RealmConfiguration.Builder.assetFile(String)`.
+* `Realm.createObject(Class<E>)` and `DynamicRealm.createObjec(String)` throws a `RealmException` if they are called to create an object with a primary key defined.
+* Importing from JSON without the primary key field defined in the JSON object now throws an exception.
```
Comment (kneth) [inline CHANGELOG.md]: which exception?

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,8 @@
   - `RealmIOExcpetion` has been removed and replaced by `RealmFileException`.
 * Removed `RealmConfiguration.Builder(Context, File)` and `RealmConfiguration.Builder(File)` constructors.
 * `RealmConfiguration.Builder.assetFile(Context, String)` has been renamed to `RealmConfiguration.Builder.assetFile(String)`.
+* `Realm.createObject(Class<E>)` and `DynamicRealm.createObjec(String)` throws a `RealmException` if they are called to create an object with a primary key defined.
```
Comment (kneth) [inline CHANGELOG.md]: "throws a" -> "throws"

File: realm/realm-library/src/androidTest/java/io/realm/RealmJsonAbsentPrimaryKeyTests.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @After
+    public void tearDown() {
+        if (realm != null) {
+            realm.close();
+        }
+    }
+
+    // parameters for testing absent primary key value. PrimaryKey field isabsent.
```
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/RealmJsonAbsentPrimaryKeyTests.java]: space missing after "is"

File: realm/realm-library/src/main/java/io/realm/internal/Table.java [hunk 4]
```diff
@@ -373,25 +373,18 @@ public void moveLastOver(long rowIndex) {
         nativeMoveLastOver(nativePtr, rowIndex);
     }
 
-    public long addEmptyRow() {
+    /**
+     * Add an empty row to the table.
+     *
+     * @param checkPrimaryKey set to true to check if this table has a primary key defined. Throw an exception if yes.
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/internal/Table.java]: Embed true in `@code`

File: realm/realm-library/src/main/java/io/realm/internal/Table.java [hunk 5]
```diff
@@ -373,25 +373,18 @@ public void moveLastOver(long rowIndex) {
...
-                    break;
-                case INTEGER:
-                    if (findFirstLong(primaryKeyColumnIndex, INTEGER_DEFAULT_VALUE) != NO_MATCH) {
-                        throwDuplicatePrimaryKeyException(INTEGER_DEFAULT_VALUE);
-                    }
-                    break;
-                default:
-                    throw new RealmException("Cannot check for duplicate rows for unsupported primary key type: " + type);
-            }
+        if (!checkPrimaryKey && hasPrimaryKey()) {
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/internal/Table.java]: In order to throw, both subexpressions must be true, right? So `checkPrimaryKey` must be false. This is not consistent with doc above. Or what am I missing :-)

Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/internal/Table.java]: Nice catch! will fix and add a test case.

**Comment by zaki50** [comment]
Shouldn't we update Javadoc comments of `Realm#create*FromJson()`?

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 6]
```diff
@@ -1616,9 +1623,16 @@ private void emitCreateUsingJsonStream(JavaWriter writer) throws IOException {
...
+        if (metadata.hasPrimaryKey()) {
+            writer.beginControlFlow("if (!jsonHasPrimaryKey)");
+            writer.emitStatement(Constants.STATEMENT_EXCEPTION_NO_PRIMARY_KEY_IN_JSON, metadata.getPrimaryKey());
+            writer.endControlFlow();
+        }
+        writer.emitStatement("obj = realm.copyToRealm(obj)");
         writer.emitStatement("return obj");
         writer.endMethod();
         writer.emitEmptyLine();
+
```
Comment (zaki50) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: remove this?

File: realm/realm-library/src/androidTest/java/io/realm/RealmJsonNullPrimaryKeyTests.java [hunk 7]
```diff
@@ -130,23 +125,28 @@ public void createOrUpdateObjectFromJson_primaryKey_isNullOrAbsent_fromJsonObjec
...
-        realm.commitTransaction();
-
+    public void createOrUpdateObjectFromJson_primaryKey_isNull_updateFromJsonObject() throws JSONException {
         // PrimaryKeyAsString
         if (clazz.equals(PrimaryKeyAsString.class)) {
+            realm.beginTransaction();
+            realm.createObject(clazz, null); // name = null, id =null
+            realm.createOrUpdateObjectFromJson(clazz, new JSONObject(jsonString));
+            realm.commitTransaction();
+
```
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmJsonNullPrimaryKeyTests.java]: Why do we have these same lines in both `if` and `else`?

Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/RealmJsonNullPrimaryKeyTests.java]: HA, right, will change it. It was caused by that i thought the default value is `0` and `""`.

File: realm/realm-library/src/androidTest/java/io/realm/internal/JNITableTest.java [hunk 8]
```diff
@@ -87,7 +87,7 @@ public void zeroColOperations() {
 
         // Add rows
         try { tableZeroCols.add("val");         fail("No columns in table"); } catch (IndexOutOfBoundsException ignored) {}
-        try { tableZeroCols.addEmptyRow();      fail("No columns in table"); } catch (IndexOutOfBoundsException ignored) {}
+        try { tableZeroCols.addEmptyRow(true);      fail("No columns in table"); } catch (IndexOutOfBoundsException ignored) {}
```
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/internal/JNITableTest.java]: please align `fail()` to other lines

File: realm/realm-library/src/main/java/io/realm/internal/Table.java [hunk 9]
```diff
@@ -373,25 +373,18 @@ public void moveLastOver(long rowIndex) {
...
+            throw new RealmException(String.format("Class '%s' has a primary key defined and the primary key needs to" +
+                    " be set when creating new object.", getName()));
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/Table.java]: `getName()` should be `Table.tableNameToClassName(getName())`

**Comment by beeender** [comment]
@realm/java Updated.

File: CHANGELOG.md [hunk 10]
```diff
@@ -9,6 +9,8 @@
   - `RealmIOExcpetion` has been removed and replaced by `RealmFileException`.
 * Removed `RealmConfiguration.Builder(Context, File)` and `RealmConfiguration.Builder(File)` constructors.
 * `RealmConfiguration.Builder.assetFile(Context, String)` has been renamed to `RealmConfiguration.Builder.assetFile(String)`.
+* `Realm.createObject(Class<E>)` and `DynamicRealm.createObjec(String)` throws `RealmException` if they are called to create an object with a primary key defined.
```
Comment (cmelchior) [inline CHANGELOG.md]: I would expand this a little I think. Something like

Objects with primary keys are now required to define it when the object is created. This means that `Realm.createObject(Class<E>)` and `DynamicRealm.createObject(String)` now throws `RealmException` if they are used to create an object with a primary key field. Use `createObject(Class, Object)` instead.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 11]
```diff
@@ -1553,6 +1553,10 @@ private void emitCreateOrUpdateUsingJsonObject(JavaWriter writer) throws IOExcep
         writer.emitEmptyLine();
     }
 
+    // FIXME: Since we need to check the PK in stream before create an object, this is now using copyToRealm instead of
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: I am not sure we can ship this? It sounds like a major regression (at least for the ones using this API). 
I'm not sure what you mean by parsing the stream twice? Currently we convert to JSONObject, but that is the same as doing `copyToRealm`, just with JSON, no?

Comment (beeender) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: That is only for `createOrUpdateAllFromJson`, for updating. by creating directly from stream, we set the properties one by one. If this is not acceptable, I would like to implement the #777 first since it doesn't seem to be too difficult.

File: realm/realm-library/src/androidTest/java/io/realm/RealmJsonAbsentPrimaryKeyTests.java [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    public void createOrUpdateObjectFromJson_primaryKey_isNullOrAbsent_fromJsonObject() throws JSONException {
+        realm.beginTransaction();
+        thrown.expect(IllegalArgumentException.class);
+        realm.createOrUpdateObjectFromJson(clazz, new JSONObject(jsonString));
+        realm.commitTransaction();
+    }
+
+    // Testing absent primary key value for createAllFromJson()
+    @Test
+    public void createALlFromJson_primaryKey_isNullOrAbsent_fromJsonObject() throws JSONException {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmJsonAbsentPrimaryKeyTests.java]: ALl -> All

Also below

File: realm/realm-library/src/androidTest/java/io/realm/RealmJsonAbsentPrimaryKeyTests.java [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    final private String jsonString;
+
+    public RealmJsonAbsentPrimaryKeyTests(Class<? extends RealmObject> clazz, String jsonString) {
+        this.jsonString = jsonString;
+        this.clazz = clazz;
+    }
+
+    // Testing absent primary key value for createObjectFromJson()
+    @Test
+    public void createObjectFromJson_primaryKey_isNullOrAbsent_fromJsonObject() throws JSONException {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmJsonAbsentPrimaryKeyTests.java]: Shouldn't it just be `isAbsent`, it isn't null in any of the test data?

Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/RealmJsonAbsentPrimaryKeyTests.java]: Uh... forget to change this name. Copied from the original test.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 11 above)
Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: It looks like it only effect the `create` part, not the `update` part right?
I see 3 choices:

1) Accept the current solution and document the breaking change 
2) Rewrite `createFromJsonStream` to create JSONObject instead and return the result from `createObjectFromJson(jsonObj)`
3) Add support for #777 

For expendiency, I would probably just go with 1) now make an issue for looking into 2/3 in th next week or two, because we need a better solution before releasing anything publically.

If 3) is indeed fairly easy to do that would be awesome, if not I think we should do 2.

Comment (beeender) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: See #3386

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 14]
```diff
@@ -675,6 +687,11 @@ private Scanner getFullStringScanner(InputStream in) {
     public <E extends RealmModel> E createObject(Class<E> clazz) {
         checkIfValid();
         Table table = schema.getTable(clazz);
+        // Check and throw the exception earlier for a better exception message.
+        if (table.hasPrimaryKey()) {
+            throw new RealmException(String.format("'%s' has a primary key, use" +
+                    " 'createObject(Class<E>, Object)' to create it.", clazz.getSimpleName()));
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `clazz.getSimpleName()` returns obfuscated name if ProGuard was applied.
`Table.tableNameToClassName(table.getName())` is better I think.

File: realm/realm-library/src/main/java/io/realm/internal/Table.java [hunk 15]
```diff
@@ -475,7 +469,7 @@ public long addEmptyRows(long rows) {
      * @return the row index of the appended row.
      */
     protected long add(Object... values) {
-        long rowIndex = addEmptyRow(false);
+        long rowIndex = addEmptyRow();
 
         checkImmutable();
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/Table.java]: Though this is not related to the changes of PR, Is it OK to check immutability after adding a row?

Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/Table.java]: @beeender How about this?
Moved some checks first and supported primary key.

```
diff --git a/realm/realm-library/src/main/java/io/realm/internal/Table.java b/realm/realm-library/src/main/java/io/realm/internal/Table.java
index 5b3c1ba..4b557b7 100644
--- a/realm/realm-library/src/main/java/io/realm/internal/Table.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/Table.java
@@ -481,8 +481,6 @@ public class Table implements TableOrView, TableSchema {
      * @return the row index of the appended row.
      */
     protected long add(Object... values) {
-        long rowIndex = addEmptyRow();
-
         checkImmutable();
 
         // Check values types
@@ -512,8 +510,14 @@ public class Table implements TableOrView, TableSchema {
             }
         }
 
+        final long pkIndex = getPrimaryKey();
+        long rowIndex = (pkIndex == NO_MATCH) ? addEmptyRow() : addEmptyRowWithPrimaryKey(values[(int) pkIndex]);
+
         // Insert values
         for (long columnIndex = 0; columnIndex < columns; columnIndex++) {
+            if (columnIndex == pkIndex) {
+                continue;
+            }
             Object value = values[(int)columnIndex];
             switch (colTypes[(int)columnIndex]) {
             case BOOLEAN:
```

**Comment by zaki50** [comment]
Since primary key check is removed from `addEmptyRow()`, I think it's better to update `io.realm.internal.Table#add(Object...)` like:

```
diff --git a/realm/realm-library/src/main/java/io/realm/internal/Table.java b/realm/realm-library/src/main/java/io/realm/internal/Table.java
index 5b3c1ba..4b557b7 100644
--- a/realm/realm-library/src/main/java/io/realm/internal/Table.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/Table.java
@@ -481,8 +481,6 @@ public class Table implements TableOrView, TableSchema {
      * @return the row index of the appended row.
      */
     protected long add(Object... values) {
-        long rowIndex = addEmptyRow();
-
         checkImmutable();
 
         // Check values types
@@ -512,8 +510,14 @@ public class Table implements TableOrView, TableSchema {
             }
         }
 
+        final long pkIndex = getPrimaryKey();
+        long rowIndex = (pkIndex == NO_MATCH) ? addEmptyRow() : addEmptyRowWithPrimaryKey(values[(int) pkIndex]);
+
         // Insert values
         for (long columnIndex = 0; columnIndex < columns; columnIndex++) {
+            if (columnIndex == pkIndex) {
+                continue;
+            }
             Object value = values[(int)columnIndex];
             switch (colTypes[(int)columnIndex]) {
             case BOOLEAN:
```

**Comment by zaki50** [comment]
:+1: other than some minor comments

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 11 above)
Comment (zaki50) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: If we can fix #2536, fixing #777 is very easy (just enabling the transforming in the constructor).

**Comment by beeender** [comment]
@zaki50 Yes, saw it. I was thinking the method should just be removed since i cannot see why it is useful. I will put some comments to the function.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 11 above)
Comment (beeender) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: I was thinking something like we do byte code insertion a special constructor there, something like `Foo(RealmDefaultConstructor c)` if user doesn't have one there. With the construction, user can set the default values.

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 14 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: @zaki50 But `clazz` here should have `@Keep` right? otherwise we cannot even get the table at line 689

Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Model classes don't have `@Keep` annotation.
Table name is resolved by `RealmProxyMediator#getTableName()` and it uses `Class` itself, not it's name.

**Comment by zaki50** [comment]
:+1:

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,9 @@
 ## 2.0.0
 
+### Known issues
+
+* When create `RealmObject` from JSON stream, it will takes default values set by the non-managed `RealmObject`'s default constructor for those absent field in JSON object. This behaviour is different from other APIs when creating `RealmObject` (#3386).
```
Comment (cmelchior) [inline CHANGELOG.md]: When creating a `RealmObject` from a JSON stream, it will take the default values defined by its default constructor for those fields that are not defined in the JSON object. This behaviour is different from other APIs when creating `RealmObject`s

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 17]
```diff
@@ -675,6 +687,11 @@ private Scanner getFullStringScanner(InputStream in) {
     public <E extends RealmModel> E createObject(Class<E> clazz) {
         checkIfValid();
         Table table = schema.getTable(clazz);
+        // Check and throw the exception earlier for a better exception message.
+        if (table.hasPrimaryKey()) {
+            throw new RealmException(String.format("'%s' has a primary key, use" +
+                    " 'createObject(Class<E>, Object)' to create it.", Table.tableNameToClassName(table.getName())));
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: to create it -> instead.

**Comment by cmelchior** [comment]
Except some minor comments and if CI is happy 👍
